### ISSUE TITLE
test_iocp's _acceptAddressTest should use port as SO_UPDATE_ACCEPT_CONTEXT similar to iocpreactor.tcp.Port

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5356]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)| @zseil reported|
|-|-|
|Trac ID|trac#5356|
|Type|defect|
|Created|2011-11-02 15:22:17Z|
|Branch|https://github.com/twisted/twisted/tree/test-ipv6AcceptAddress-5356|

The `server.setsockopt()` in `test_iocp.SupportTests._acceptAddressTest` should use the listening  `port` socket as the sockopt value, not the accepted `server` socket. That way it would match what `iocpreactor.tcp.Port` does and what msdn documentation for AcceptEx says:
http://msdn.microsoft.com/en-us/library/ms737524%28v=VS.85%29.aspx#CodeSnippetContainerCode_ff3101dc-1681-493d-9d74-e2551a72bb0b

(it looks like that is the only msdn page where documentation for SO_UPDATE_ACCEPT_CONTEXT matches reality)

I think this bug might be the cause of this error:
```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "\twisted\internet\test\test_iocp.py", line 82, in test_ipv4AcceptAddress
    self._acceptAddressTest(AF_INET, '127.0.0.1')
  File "\twisted\internet\test\test_iocp.py", line 64, in _acceptAddressTest
    SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT, pack('I', server.fileno()))
  File "c:\python27\lib\socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 10057] A request to send or receive data was disallowed ...

twisted.internet.test.test_iocp.SupportTests.test_ipv4AcceptAddress
-------------------------------------------------------------------------------
```

seen here:
http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/571

<details><summary>Searchable metadata</summary>

```
trac-id__5356 5356
type__defect defect
reporter__zseil zseil
priority__normal normal
milestone__ 
branch__branches_test_ipv6AcceptAddress_5356 branches/test-ipv6AcceptAddress-5356
branch_author__adiroiban adiroiban
status__new new
resolution__None None
component__core core
keywords__ 
time__1320247337000000 1320247337000000
changetime__1602079213459747 1602079213459747
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Any insight as to why this only fails sometimes?

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_2"></a>|@zseil commented|
|-|-|

Bah, my guess was wrong, the test still fails intermittently after changing sockopt value to `pack('P', port.fileno())` . I was able to reproduce the error by running `SupportTests` repeatedly, it usually fails after about 1000 passes, sometimes sooner. I thought that the leak of the overlapped structure in `_iocp.accept` might be the problem (the leak is not present in IOCPReactor since it regularly calls `CompletionPort.getEvent()`), but the same error can also be reproduced with `win32file.AcceptEx()`. Even waiting for the overlapped operation to finish (by calling `GetOverlappedResult(..., wait=True)` to win32file version of the test) the test still fails. So it looks like this is a bug in Windows.

One workaround for the test failure would be to try setting the `sockopt` multiple times -- the error seems harmless, a second `setsockopt()` call always succeeded in my testing.

Should the `sockopt` value still be changed or should it be left as is? It doesn't seem to make any difference.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> Should the sockopt value still be changed or should it be left as is? It doesn't seem to make any difference.

I guess we should pass the value that seems correct, based on our best understanding of the available documentation, even if it doesn't appear to make a difference (or we should always pass 0 or some other nonsense value, making it clear to readers that that value is not important).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[#7756](https://github.com/twisted/twisted/issues/7756) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

> One workaround for the test failure would be to try setting the sockopt multiple times -- the error seems harmless, a second setsockopt() call always succeeded in my testing.

I also tried to look at this but found not solution.

+1 on this hack. We can start with re-trying one.

It should be better than requiring developers to manually re-run the test just to check that this is not a permanent failure.

What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

As long as the hack is documented that seems reasonable.

I do wonder how reliable that "always" is.  If the failure is a 1 in 1000 thing then it would presumably take a million test case runs before you might expect a good chance of seeing the call fail twice in a row...

Making this spurious failure happen one time in a million is probably still an improve here, anyway.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

(In [43882]) Branching to test-ipv6AcceptAddress-5356.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban set owner to @adiroiban<br>@adiroiban set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban removed owner<br>@adiroiban set status to `new`|
|-|-|


1. I have updated the code to use the listening port in sockopt to match the code used in iocpreactor.tcp ... but test are still failing as the socket is not connected yet so any value will fail.

2. I need to put a sleep as otherwise even if I retry 20 times tests are still failing.

3. With a wait of 0.001 and a single retry I see 1 or 2 failures out of 1000 tries.

4. Retrying 3 times further reduce the failure rate.

5. For the final patch I went for 3 retried and 0.1 second of wait.

6. I have created a follow ticket [#7818](https://github.com/twisted/twisted/issues/7818) to look at a functional test for the IOCP reactor

7. If I run the tests 10000 I see about 9000 failures... so I guest that if you try to listen to many ports Windows will either reject or use a huge delay in retrieving those ports.


Sending latest commit to all builders.

Please review. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @adiroiban|
|-|-|

Hi adi!

Thanks for working on this.  Windows maintenance contributions are _especially_ valuable since not a lot of folks follow this aspect of Twisted :).

I'm having some trouble following what this ticket is actually about though.  As I understand it, the issue being fixed here is something like "`twisted.internet.test.test_iocp.SupportTests.test_ipv4AcceptAddress` fails intermittently"?  And the patch attached improves the situation, but doesn't fully resolve it?

Here's some review feedback, all optional; you may want to submit this for re-review, but given the state of the existing code I think this patch is an improvement:

  1. Looking at the previous discussion, it looks like one way to make success more likely would be to use `GetOverlappedResult(..., wait=True)`.  Microsoft _does_ tend to fix issues like this in Windows (our XP builder has had several more it's-a-bug-in-Windows issues than our Win7 builder), so we should make it do the "right" thing first.  I think this may also be the reason that large numbers of calls fail; although the `close` is called on cleanup, there might still be an outstanding pile of overlapped results pending someplace.
  2. The comment in `handleAccept` - "Inherit properties from the listening port socket as documented" - leaves something to be desired.  "Inherit properties" - what properties? "as documented" - documented by whom?  The given link just documents `AcceptEx`, a function whose name doesn't appear anywhere near the comment because we're in a callback for a callback for it.  It would be nice to be more specific about this, relating it to some claim Twisted makes about sockets, or some much more specific behavior we wish to ensure.
  3. [10057 is WSAENOTCONN](https://msdn.microsoft.com/en-us/library/windows/desktop/ms740668(v=vs.85).aspx#wsaenotconn) and should be referenced symbolically, not numerically; you can import it from the `errno` module on Windows, as many other modules in Twisted do.

If you land this to close this ticket though, please be sure to open any follow-on issues (such as making it more reliable or improving the documentation).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|



Thanks for the review.

This ticket updates the test code to use the same code as the iocpreactor.tcp.Port and it retries on failure.

I have updated the description of the ticket.

1.  This patch does not resolve the issue with IOCP reactor. I have created ticket
    [#7839](https://github.com/twisted/twisted/issues/7839) for that.

    I had a hard time building iocp support on my XP WM so for now I have only
    fixed the test and try to gather some info for it.

    I plan to continue working on this once I get Twisted extensions available
    on my XP WM. I came from Unix background but would like to learn more
    about windows API

2.  I have extended the comment. I think that this link and current comment
    while not perfect, is better than nothing.
    I have no idea what properties are inherited. I just tried to add that
    text to make it easier to search the linked documentation.

3.  I have updated the code to use the ENOTCONN constant from twisted.internet.tcp
    Not sure if this is a good idea.
    I could not find many other modules importing WSAENOTCONN from errno.

Sending this back to review and to buildbots.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

I have no idea why this fails on buildslaves as it pass on my Ubuntu system

```
from twisted.internet.tcp import ENOTCONN
```

I have change the code to implicit import.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for this work, adi.  I have to confess though, I still don't know what this work is intended to _do_.  I still have a number of questions:

  1. If this isn't to fix the reliability issue, then what is the point of using the same code?  Why aren't you using `GetOverlappedResult(..., wait=True)` as I asked before?  In iocpreactor we don't need the retry because we only set the socket option in `handleAccept` when the full overlapped operation has already completed.
  2. You said that the goal here is to use "the same code as the iocpreactor.tcp.Port", but the code is duplicated, not simply called.  Given that the `struct.pack()` is somewhat magical, it would be good to at least put this into a shared function that is called in both places.

Regarding `WSAENOTCONN`, nobody else is importing it from errno because nobody else has needed to check for it yet (generally because when overlapped operations have completed, the connection is already established and there's no case where that error is raised).

There are some merge conflicts, but I have some simpler points to address while you're merging forward, mostly in the area of eliminating magic numbers:

 1. `3` is a magic number.  This should be a constant, `_SO_UPDATE_ACCEPT_CONTEXT_RETRY_COUNT`, let's say, with an accompanying `@var` so there's some clue as to why we're retrying 3 times.
 2. As I said before, `Errno 10057` is `WSAENOTCONN`, and should be described as such in the comment as well as the change you made to the code.
 3. The `0.1` argument to `time.sleep` is also a magic number; `_SO_UPDATE_ACCEPT_CONTEXT_RETRY_TIME`

Please address the above in some capacity and resubmit.

**Comment by twisted-trac** [comment]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)<a name="note_14"></a>|@mthuurne commented|
|-|-|

Passing the `port` socket instead of the `server` socket was implemented in [#8780](https://github.com/twisted/twisted/issues/8780).

The test case is still flaky though. I think the underlying problem is that `_acceptAddressTest()` ignores the fact that `accept()` can return while the underlying `AcceptEx()` call is still pending (`ERROR_IO_PENDING`). In `tcp.Port`, an event is passed to `accept()` and the socket options are set on the event's callback. In `_acceptAddressTest()`, `None` is passed instead of an event and no attempt is made to determine whether the underlying accept call is complete or pending; it just tries to set the socket options immediately.

In other words, the test treats `accept()` as if it blocks, but `accept()` is implemented asynchronously instead.

Possible solutions:
- use an event callback in the test case, similar to `tcp.Port`'s implementation
- retry the `setsockopt()` call a few times with short delays between attempts
- drop this low-level test and instead rely on the high-level tests in `test_tcp.py`

I don't have a Windows machine, so I won't be implementing a fix for this. But as a user of Twisted's CI, I am interested in seeing this flaky test fixed or removed.
